src/pveclib/vec_int128_ppc.h

```diff
@@ -2060,6 +2060,122 @@ static inline vui128_t vec_vmuleud (vui64_t a, vui64_t b);
 static inline vui128_t vec_vmuloud (vui64_t a, vui64_t b);
 ///@endcond
 
+/** \brief Transfer a vector unsigned __int128 to __int128 scalar.
+*
+*  The compiler does not allow direct transfer (assignment or type
+*  cast) between __int128 scalars and vector types. Vectors are held
+*  in 128-bit VRs (VSRs) and __int128 scalars are held in pair of
+*  64-bit GPRs. So this operation requires a transfer between
+*  registers of different types/sizes.
+*
+*  |processor|Latency|Throughput|
+*  |--------:|:-----:|:---------|
+*  |power8   |  6-7  | 1/cycle  |
+*  |power9   |  5-6  | 2/cycle  |
+*
+*  @param vra a vector unsigned __int128 value.
+*  @return The original value returned as a __int128 scalar.
+*/
+static inline unsigned __int128
+vec_xfer_vui128t_2_uint128 (vui128_t vra)
+{
+  __VEC_U_128 t;
+  unsigned __int128 result;
+#if defined(_ARCH_PWR8) || defined (__clang__)
+  // PWR8/9 should generate Move From VSR Doubleword instructions.
+  t.vx1 = vra;
+  result = t.ui128;
+#else
+  /* PWR7 and earlier must transfer through storage.  This requires
+   * care as we want to avoid load-hit-store flushes in the pipeline.
+   * First split the vector into a pair of dword FPRs (vra_u, vra_l). */
+  vui64_t vra_u = (vui64_t) vra;
+  vui64_t vra_l = vec_swapd ((vui64_t)vra);
+  /* Store this pair as adjacent dwords, followed by a group ending
+   * nop. This prevents the hardware from dispatching the stores in the
+   * same cycle as the following loads (a guaranteed pipeline flush).
+   * Also the load addresses and data size will match these stores and
+   * increase the possibility of store forwarding from the store queue.
+   */
+  __asm__(
+      "stfd %2,%0;"
+      "stfd %3,%1;"
+      "ori  2,2,0;"
+      : "=m" (t.ulong.lower),
+	"=m" (t.ulong.upper)
+      : "d" (vra_l), "d" (vra_u)
+      : );
+  // Load the dwords into a pair of GPRs for the __int128 result.
+  result = t.ui128;
+#endif
+  return (result);
+}
+
+/** \brief Transfer a __int128 scalar to vector unsigned __int128.
+*
+*  The compiler does not allow direct transfer (assignment or type
+*  cast) between __int128 scalars and vector types. Vectors are held
+*  in 128-bit VRs (VSRs) and __int128 scalars are held in pair of
+*  64-bit GPRs. So this operation requires a transfer between
+*  registers of different types/sizes.
+*
+*  |processor|Latency|Throughput|
+*  |--------:|:-----:|:---------|
+*  |power8   |   7   | 1/cycle  |
+*  |power9   |   5   | 1/cycle  |
+*
+*  @param grpr a unsigned __int128 value.
+*  @return The original value returned as a vector unsigned__int128.
+*/
+static inline vui128_t
+vec_xfer_uint128_2_vui128t (unsigned __int128 gprp)
+{
+  __VEC_U_128 t;
+  t.ui128 = gprp;
+  return t.vx1;
+}
+
+/** \brief Extract the low doubleword from a __int128 scalar.
+*
+*  @param grpr a unsigned __int128 value.
+*  @return The low doubleword of __int128.
+*/
+static inline unsigned long long
+ext_uint128_low (unsigned __int128 gprp)
```

Comment:
No, __int128 is a scalar implemented a a GPR pair (GPRp). This is a separate type from vector _int128 which is a vector. 

If you want a a prefix it would be scalar_ (as in scalar_insert_exp and scalar_test_data_class). See: https://gcc.gnu.org/onlinedocs/gcc-10.2.0/gcc/PowerPC-AltiVec-Built-in-Functions-Available-on-ISA-3_002e0.html#PowerPC-AltiVec-Built-in-Functions-Available-on-ISA-3_002e0

Also extract (ext) and Insert (ins) are distinct from transfer (xfer). GCC spells out extract/insert while I prefer the shorted forms.

What is your preference?

Reply:
Since we already infringe on `scalar`, then I'd be content with `scalar_extract_uint64_from_uint128l`, which seems to match closest to GCC's roughly similar `vec_extract_fp32_from_shortl`.  I think we should keep it confined to an existing namespace (`scalar`). Whether you use `extract` or `ext` is personal preference, although I read `ext` as "extension" generally. Using the `uint` types seems more consistent with the rest of the API.